testing/btest/language/vector-set-conversions.zeek

```diff
@@ -0,0 +1,24 @@
+# @TEST-DOC: Tests converting between sets and vectors with the 'as' keyword
+# @TEST-EXEC: zeek -b %INPUT > out
+# @TEST-EXEC: btest-diff out
+# @TEST-EXEC: btest-diff .stderr
+
+print("vector to set");
+local v1 = vector(1, 1, 1, 2, 2, 3, 3, 4);
+local s1 = v1 as set[count];
+print(v1);
+print(s1);
+
+print("");
+print("set to vector (count)");
+local s2 = set(1, 2, 3, 4);
+local v2 = s2 as vector of count;
+print(s2);
+print(v2);
+
+print("");
+print("set to vector (port)");
+local s3 = set(21/tcp, 23/tcp);
+local v3 = s3 as vector of port;
+print(s3);
+print(v3);
```

Comment:
I've tried the following examples with nested records/containers and they all trigger ASAN errors in `TypeList::Append()`. Probably avoided by re-using `Type t` - haven't looked what's actually happening.

Record types also bring up the question of coercing, but agree to ignore for now and keep conversion restricted to identical yield types.

```
# t1.zeek
type r: record {
        a: string;
        b: string;
};

global s: set[r] = set([$a="a", $b="b"], [$a="a1", $b="b1"]);
global v: vector of r = vector([$a="a", $b="b"], [$a="a1", $b="b1"]);

print s;
print v;
print s as vector of r;
print v as set[r];
```
```
# t2.zeek
global s: set[vector of count] = set(vector(1), vector(2));
global v: vector of vector of count = vector(vector(3), vector(4));

print s;
print v;
print s as vector of vector of count;
print v as set[vector of count];
```
```
# t3.zeek

global s: set[set[count]] = set(set(1), set(2));
global v: vector of set[count] = vector(set(3), set(4));

print s;
print v;
print s as vector of set[count];
print v as set[set[count]];
```


```
$ zeek t1.zeek 
{
[a=a, b=b],
[a=a1, b=b1]
}
[[a=a, b=b], [a=a1, b=b1]]
=================================================================
==1125401==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x606000c5c3f8 at pc 0x55cdf10aa575 bp 0x7ffc77fea1a0 sp 0x7ffc77fea198
READ of size 4 at 0x606000c5c3f8 thread T0
    #0 0x55cdf10aa574 in zeek::RecordType::NumFields() const /home/awelzel/corelight-oss/zeek/src/include/zeek/Type.h:702
    #1 0x55cdf10aa574 in zeek::same_type(zeek::Type const&, zeek::Type const&, bool, bool) /home/awelzel/corelight-oss/zeek/src/Type.cc:2118
    #2 0x55cdf10abdc5 in zeek::same_type(zeek::IntrusivePtr<zeek::Type> const&, zeek::IntrusivePtr<zeek::Type> const&, bool, bool) /home/awelzel/corelight-oss/zeek/src/include/zeek/Type.h:919
    #3 0x55cdf10abdc5 in zeek::TypeList::Append(zeek::IntrusivePtr<zeek::Type>) /home/awelzel/corelight-oss/zeek/src/Type.cc:301
    #4 0x55cdf10f55a1 in zeek::ListVal::Append(zeek::IntrusivePtr<zeek::Val>) /home/awelzel/corelight-oss/zeek/src/Val.cc:1596
    #5 0x55cdf11068ce in zeek::TableVal::ToListVal(zeek::TypeTag) const /home/awelzel/corelight-oss/zeek/src/Val.cc:2602
    #6 0x55cdf110748b in zeek::TableVal::ToPureListVal() const /home/awelzel/corelight-oss/zeek/src/Val.cc:2618
    #7 0x55cdf112c86b in zeek::cast_value_to_type(zeek::Val*, zeek::Type*) /home/awelzel/corelight-oss/zeek/src/Val.cc:4277
    #8 0x55cdf0d510bb in zeek::detail::cast_value(zeek::IntrusivePtr<zeek::Val>, zeek::IntrusivePtr<zeek::Type> const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) /home/awelzel/corelight-oss/zeek/src/Expr.cc:5236
    #9 0x55cdf0d69d0b in zeek::detail::CastExpr::Fold(zeek::Val*) const /home/awelzel/corelight-oss/zeek/src/Expr.cc:5226
    #10 0x55cdf0d52b12 in zeek::detail::UnaryExpr::Eval(zeek::detail::Frame*) const /home/awelzel/corelight-oss/zeek/src/Expr.cc:652
    #11 0x55cdf0d702f5 in zeek::detail::eval_list(zeek::detail::Frame*, zeek::detail::ListExpr const*) /home/awelzel/corelight-oss/zeek/src/Expr.cc:5506
    #12 0x55cdf1044ae0 in zeek::detail::ExprListStmt::Exec(zeek::detail::Frame*, zeek::detail::StmtFlowType&) /home/awelzel/corelight-oss/zeek/src/Stmt.cc:274
    #13 0x55cdf1046de2 in zeek::detail::StmtList::Exec(zeek::detail::Frame*, zeek::detail::StmtFlowType&) /home/awelzel/corelight-oss/zeek/src/Stmt.cc:1688
    #14 0x55cdf0b9c4e2 in zeek::detail::setup(int, char**, zeek::Options*) /home/awelzel/corelight-oss/zeek/src/zeek-setup.cc:1107
    #15 0x55cdf1868916 in main /home/awelzel/corelight-oss/zeek/src/main.cc:57
    #16 0x7f05d2cda1c9 in __libc_start_call_main ../sysdeps/nptl/libc_start_call_main.h:58
    #17 0x7f05d2cda284 in __libc_start_main_impl ../csu/libc-start.c:360
    #18 0x55cdefb3e760 in _start (/home/awelzel/corelight-oss/zeek/build/src/zeek+0x9f8760)
```

Reply:
> Probably avoided by re-using Type t - haven't looked what's actually happening.

That's not it, as I fixed that comment and it's still doing it. Somewhere along the lines of the call to `ToPureListVal` the base type pointed to for `TYPE_RECORD` gets corrupted (deleted?) and so `same_type` fails. You can actually more easily see it by building with the undefined behavior analyzer:

```
/Users/tim/Desktop/projects/zeek/src/Type.cc:116:9: runtime error: downcast of address 0x000113115140 which does not point to an object of type 'const zeek::RecordType'
0x000113115140: note: object is of type 'zeek::Type'
 00 00 00 00  10 cf b6 05 01 00 00 00  50 d5 28 13 01 00 00 00  02 00 00 00 00 be be be  0f 00 00 00
              ^~~~~~~~~~~~~~~~~~~~~~~
              vptr for 'zeek::Type'
    #0 0x10334016c in zeek::Type::AsRecordType() const Type.cc:116
    #1 0x10337f3ec in zeek::same_type(zeek::Type const&, zeek::Type const&, bool, bool) Type.cc:2116
    #2 0x1017c0a4c in zeek::same_type(zeek::IntrusivePtr<zeek::Type> const&, zeek::IntrusivePtr<zeek::Type> const&, bool, bool) Type.h:919
    #3 0x103345ec4 in zeek::TypeList::Append(zeek::IntrusivePtr<zeek::Type>) Type.cc:301
    #4 0x10343fd48 in zeek::ListVal::Append(zeek::IntrusivePtr<zeek::Val>) Val.cc:1596
    #5 0x103466574 in zeek::TableVal::ToListVal(zeek::TypeTag) const Val.cc:2602
    #6 0x103466cf4 in zeek::TableVal::ToPureListVal() const Val.cc:2618
    #7 0x103498600 in zeek::cast_value_to_type(zeek::Val*, zeek::Type*) Val.cc:4277
    #8 0x102cece98 in zeek::detail::cast_value(zeek::IntrusivePtr<zeek::Val>, zeek::IntrusivePtr<zeek::Type> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&) Expr.cc:5236
    #9 0x102cec970 in zeek::detail::CastExpr::Fold(zeek::Val*) const Expr.cc:5226
    #10 0x102c40178 in zeek::detail::UnaryExpr::Eval(zeek::detail::Frame*) const Expr.cc:652
    #11 0x102cc890c in zeek::detail::eval_list(zeek::detail::Frame*, zeek::detail::ListExpr const*) Expr.cc:5506
    #12 0x103278d2c in zeek::detail::ExprListStmt::Exec(zeek::detail::Frame*, zeek::detail::StmtFlowType&) Stmt.cc:274
    #13 0x1032a8b74 in zeek::detail::StmtList::Exec(zeek::detail::Frame*, zeek::detail::StmtFlowType&) Stmt.cc:1688
    #14 0x1027dc988 in zeek::detail::setup(int, char**, zeek::Options*) zeek-setup.cc:1107
    #15 0x104433660 in main main.cc:57
    #16 0x18a677f24  (<unknown module>)
```

I'm not sure what's actually triggering that yet. If you leave the `print` statements in there, asan says that `PrintExpr` is what causes the deletion. If you take those out and just have the set-to-vector cast, asan says it's `UsageAnalyzer` which makes very little sense:

```
==53413==ERROR: AddressSanitizer: heap-use-after-free on address 0x00012100bbc0 at pc 0x0001037d8080 bp 0x00016d924510 sp 0x00016d924508
READ of size 4 at 0x00012100bbc0 thread T0
    #0 0x1037d807c in zeek::same_type(zeek::Type const&, zeek::Type const&, bool, bool) Type.cc:2118
    #1 0x1037b87f4 in zeek::TypeList::Append(zeek::IntrusivePtr<zeek::Type>) Type.cc:301
    #2 0x10380b664 in zeek::ListVal::Append(zeek::IntrusivePtr<zeek::Val>) Val.cc:1596
    #3 0x103823720 in zeek::TableVal::ToListVal(zeek::TypeTag) const Val.cc:2602
    #4 0x103824158 in zeek::TableVal::ToPureListVal() const Val.cc:2618
    #5 0x1038401f4 in zeek::cast_value_to_type(zeek::Val*, zeek::Type*) Val.cc:4277
    #6 0x10356fa64 in zeek::detail::cast_value(zeek::IntrusivePtr<zeek::Val>, zeek::IntrusivePtr<zeek::Type> const&, std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char>>&) Expr.cc:5236
    #7 0x10356f5dc in zeek::detail::CastExpr::Fold(zeek::Val*) const Expr.cc:5226
    #8 0x103510954 in zeek::detail::UnaryExpr::Eval(zeek::detail::Frame*) const Expr.cc:652
    #9 0x10355d094 in zeek::detail::eval_list(zeek::detail::Frame*, zeek::detail::ListExpr const*) Expr.cc:5506
    #10 0x103773bf0 in zeek::detail::ExprListStmt::Exec(zeek::detail::Frame*, zeek::detail::StmtFlowType&) Stmt.cc:274
    #11 0x103788afc in zeek::detail::StmtList::Exec(zeek::detail::Frame*, zeek::detail::StmtFlowType&) Stmt.cc:1688
    #12 0x1033c0c24 in zeek::detail::setup(int, char**, zeek::Options*) zeek-setup.cc:1107
    #13 0x103d73c48 in main main.cc:57
    #14 0x18a677f24  (<unknown module>)

0x00012100bbc0 is located 32 bytes inside of 64-byte region [0x00012100bba0,0x00012100bbe0)
freed by thread T0 here:
    #0 0x105b5d7b0 in wrap__ZdlPv+0x74 (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x517b0) (BuildId: f0a7ac5c49bc3abc851181b6f92b308a32000000200000000100000000000b00)
    #1 0x103bc916c in zeek::detail::UsageAnalyzer::FindSeeds(std::__1::unordered_set<zeek::detail::ID const*, std::__1::hash<zeek::detail::ID const*>, std::__1::equal_to<zeek::detail::ID const*>, std::__1::allocator<zeek::detail::ID const*>>&) const UsageAnalyzer.cc:117
    #2 0x103bc7ac4 in zeek::detail::UsageAnalyzer::UsageAnalyzer(std::__1::vector<zeek::detail::FuncInfo, std::__1::allocator<zeek::detail::FuncInfo>>&) UsageAnalyzer.cc:35
    #3 0x103bc9c50 in zeek::detail::UsageAnalyzer::UsageAnalyzer(std::__1::vector<zeek::detail::FuncInfo, std::__1::allocator<zeek::detail::FuncInfo>>&) UsageAnalyzer.cc:22
    #4 0x103b8e810 in zeek::detail::analyze_scripts(bool) ScriptOpt.cc:562
    #5 0x1033c0444 in zeek::detail::setup(int, char**, zeek::Options*) zeek-setup.cc:994
    #6 0x103d73c48 in main main.cc:57
    #7 0x18a677f24  (<unknown module>)

previously allocated by thread T0 here:
    #0 0x105b5d370 in wrap__Znwm+0x74 (libclang_rt.asan_osx_dynamic.dylib:arm64e+0x51370) (BuildId: f0a7ac5c49bc3abc851181b6f92b308a32000000200000000100000000000b00)
    #1 0x103bc9200 in zeek::detail::UsageAnalyzer::FindSeeds(std::__1::unordered_set<zeek::detail::ID const*, std::__1::hash<zeek::detail::ID const*>, std::__1::equal_to<zeek::detail::ID const*>, std::__1::allocator<zeek::detail::ID const*>>&) const UsageAnalyzer.cc:117
    #2 0x103bc7ac4 in zeek::detail::UsageAnalyzer::UsageAnalyzer(std::__1::vector<zeek::detail::FuncInfo, std::__1::allocator<zeek::detail::FuncInfo>>&) UsageAnalyzer.cc:35
    #3 0x103bc9c50 in zeek::detail::UsageAnalyzer::UsageAnalyzer(std::__1::vector<zeek::detail::FuncInfo, std::__1::allocator<zeek::detail::FuncInfo>>&) UsageAnalyzer.cc:22
    #4 0x103b8e810 in zeek::detail::analyze_scripts(bool) ScriptOpt.cc:562
    #5 0x1033c0444 in zeek::detail::setup(int, char**, zeek::Options*) zeek-setup.cc:994
    #6 0x103d73c48 in main main.cc:57
    #7 0x18a677f24  (<unknown module>)
```

Reply:
Fixed it. `TableVal::ToPureListVal()` doesn't function correctly if the index type is a record, due to the `Type` object returned by `base_type(TYPE_RECORD)` failing a typecast later on (which is what ubsan complains about). I removed the use of that and switched over to pulling the index values manually. I'll open another ticket about giving a nicer error if you try to use `ToPureListVal()` in that fashion.